Create LocalFolder explainer.md (#761)
MicrosoftEdge/MSEdgeExplainers: LocalFolder/explainer.md

The proposed solution works by making use of the [Origin Private File System (OPFS)](https://web.dev/articles/origin-private-file-system) and the [File System Access API](https://wicg.github.io/file-system-access/). It exposes an app's LocalFolder file system directory as an entry in the app origin's OPFS root directory.

## Goals

* Enable a Microsoft Store PWA to:
   * read files from the `LocalFolder` directory specific to its PFN.
   * delete any file and sub-directory in its `LocalFolder` directory. 
* Avoid prompting end users to provide input or confirmation.
* Avoid naming collisions with other OPFS root entries. 

## Non-goals

* Enable a Microsoft Store PWA to create new files or modify existing files in its LocalFolder directory.
* Support web apps not installed and managed by the Microsoft Store.
* Support web apps on non-Windows platforms.
* Standardization as a web specification.
* Introduce a non-standard Web API

## Proposed Solution 

### Existing behavior

> `navigator.storage.getDirectory()` is an existing API that returns a promise to a FileSystemDirectoryHandle representing the root of the origin's OPFS storage space. By default, an origin's OPFS root directory will have no entries.

Reviewer: nit: Would be helpful to link to have these existing API's directly link to their corresponding documentation.

```suggestion
[`navigator.storage.getDirectory()`](https://developer.mozilla.org/en-US/docs/Web/API/StorageManager/getDirectory) is an existing API that returns a promise to a [`FileSystemDirectoryHandle`] (https://developer.mozilla.org/en-US/docs/Web/API/FileSystemDirectoryHandle) representing the root of the origin's OPFS storage space. By default, an origin's OPFS root directory will have no entries.
```

> ### New behavior

Author: These directories do not accrue storage space to the origin and so the result of `navigator.storage.estimate()` will not change whether they are included in the entries under OPFS root. 

These folders do take space on disk and if cleared, can affect the results of `navigator.storage.estimate()`.

Will details add to explainer.

> Within a Microsoft Store PWA context, the directory handle for OPFS root will contain an additional entry that represents the app's LocalFolder location. This entry can be retrieved as a `FileSystemDirectoryHandle` by calling `.entries()` or `.getDirectory(...)` on the OPFS root directory handle.

Reviewer: ```suggestion
Within a Microsoft Store PWA context, the directory handle for OPFS root will contain an additional entry that represents the app's LocalFolder location. This entry can be retrieved as a [`FileSystemDirectoryHandle`](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemDirectoryHandle) by calling [`.entries()`](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemDirectoryHandle/entries) or [`.getDirectoryHandle(...)`](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemDirectoryHandle/getDirectoryHandle) on the OPFS root directory handle.

```

The LocalFolder directory handle and its contents will have properties different from other OPFS directory handles - they will only have read and delete permissions instead of read and write permissions.

```JS

>     // .getDirectory()

Reviewer: ```suggestion
    // .getDirectoryHandle()
```

    let opfsRoot = await navigator.storage.getDirectory();
    if (opfsRoot) {

>         let localStorage = await opfsRoot.getDirectoryHandle("microsoft_store_app_local_folder_APPID.37853FC22B2CE_6rarf9sa4v8jt", {create: false}); 

Reviewer: You're explicitly calling out the optional parameter of `{create: false}` here.  What would happen if you passed in `{create: true}'?  We should document that (or at least have a comment in this code snippet to address that).

Author: Added note.

> The LocalFolder directory handle and its contents will have properties different from other OPFS directory handles - they will only have read and delete permissions instead of read and write permissions.

Reviewer 2: Nit: There's either `"read"` or `"readwrite"`. There's no special delete-only case.

Author: I'll rephrase. The handle and its contents will have `read` permission without the ability to elevate to `readwrite` permission. The exception to this is when [`remove`](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemHandle/remove) or [`removeEntry`](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemDirectoryHandle/removeEntry) is used. With respect to those methods, the handles will have `readwrite` permission.

>         let localStorage = await opfsRoot.getDirectoryHandle("microsoft_store_app_local_folder_APPID.37853FC22B2CE_6rarf9sa4v8jt", {create: false}); 

Reviewer 2: ```suggestion
        let localFolder = await opfsRoot.getDirectoryHandle("microsoft_store_app_local_folder_APPID.37853FC22B2CE_6rarf9sa4v8jt", {create: false}); 
```

```

```JS
    // .entries()
    let opfsRoot = await navigator.storage.getDirectory();
    if (opfsRoot) {
        for await (const [key, value] of opfsRoot.entries()) {
            if (key == `microsoft_store_app_local_folder_APPID.37853FC22B2CE_6rarf9sa4v8jt`) {
                // ...
            }
        }
    }
```

### Deleting files

> LocalStorage contents can be cleared by calling `.removeEntry(...)` from the OPFS root directory handle.

Reviewer: ```suggestion
LocalStorage contents can be cleared by calling `.removeEntry(...)` from the OPFS root directory handle.  While you cannot delete the entry itself, you can delete all of its contents.
```

>             if (key == `microsoft_store_app_local_folder_APPID.37853FC22B2CE_6rarf9sa4v8jt`) {

Reviewer 2: ```suggestion
            if (key === 'microsoft_store_app_local_folder_APPID.37853FC22B2CE_6rarf9sa4v8jt') {
```

```JS
    let opfsRoot = await navigator.storage.getDirectory();
    if (opfsRoot) {
        await opfsRoot.removeEntry("microsoft_store_app_local_folder_APPID.37853FC22B2CE_6rarf9sa4v8jt", { recursive: true }); 
    }
```

> Alternatively, contents in LocalFolder can be deleted by calling `remove()` on file and directory handles directory.

Reviewer: ```suggestion
Alternatively, contents in LocalFolder can be deleted by calling `remove()` on file and directory handles directory.  Again, the entry itself will not be deleted, but its contents will be.
```

```JS
    let localStorage = await opfsRoot.getDirectoryHandle("microsoft_store_app_local_folder_APPID.37853FC22B2CE_6rarf9sa4v8jt", {create: false}); 

    for await (const [key, value] of localStorage.entries()) {
        value.remove();
    }
```

> ### Entry name

Reviewer: I'd probably move this section above the code examples so that it's more clear where that name is coming from when reading them.

Author: Moved "Entry name" section to above code examples.

> ### Entry name

Reviewer: We should reiterate in this section that:

1. the entry will only exist for PWA's installed through the Microsoft Store, and even then, only when the OPFS API is being used within the context of the PWA.
2. that it won't be accessible through background web workers
3. that it will exist for Microsoft Store PWA's, even if the LocalFolder itself is empty.
4. that the entry can't be deleted, but that its internal contents _can_ be deleted

Author: 1 and 2 are stated in the "Microsoft Store PWA context" section. 

Re. 3. I think it's ok for the entry to not exist for MS Store PWAs if for some reason the directory is not found on the file system. 

4. I'm reconsidering this requirement.

Author: Updated deletion behavior.

> The LocalFolder entry within the OPFS root directory can be found under the name `microsoft_store_app_local_folder{PFN}` where `{PFN}` is the Package Family Name of the app.

Reviewer: This naming is inconsistent with the examples above (missing an underscore before the `{PFN}`:

```suggestion
The LocalFolder entry within the OPFS root directory can be found under the name `microsoft_store_app_local_folder_{PFN}` where `{PFN}` is the Package Family Name of the app.
```

### Microsoft Store PWA context

To be in a Microsoft Store app context, `navigator.storage.getDirectory()` must be called from:
* Within a top level document
* Within a web app window of a Microsoft Store installed PWA
* From the main thread and not a Web Worker

### Avoiding collisions

> To avoid collisions, the LocalFolder entry under the OPFS root directory is assigned a specific name that contains the Package Family Name. If an entry with the same name already exists under the OPFS root directory, an entry representing the LocalFolder directory will not be made visible until the existing entry is renamed or otherwise removed.

Reviewer: Would be good to also link to some documentation that explains what a PFN is and how to get it.  I found [this one article](https://learn.microsoft.com/en-us/mem/configmgr/protect/deploy-use/find-a-pfn-for-per-app-vpn) on an easy way to find it, but there may be better ones.  While the article is not specific to VPN's, it's under a VPN topic, so it might be better to find a different one.

Author: Added link above in first reference to PFN.

## Similar storage folders

> There are other UWP storage APIs similar to LocalFolder that can be exposed the same way through OPFS directory handles. 

Reviewer: ```suggestion
There are other UWP storage APIs similar to LocalFolder that can be exposed the same way through OPFS directory handles, but are not currently planned to be supported at this time.
```

* [LocalCacheFolder](https://learn.microsoft.com/en-us/uwp/api/windows.storage.applicationdata.localfolder?view=winrt-22621)
* [RoamingFolder](https://learn.microsoft.com/en-us/uwp/api/windows.storage.applicationdata.roamingfolder?view=winrt-22621)
* [SharedLocalFolder](https://learn.microsoft.com/en-us/uwp/api/windows.storage.applicationdata.sharedlocalfolder?view=winrt-22621)
* [TemporaryFolder](https://learn.microsoft.com/en-us/uwp/api/windows.storage.applicationdata.temporaryfolder?view=winrt-22621)

> ## Alternatives considered

Reviewer: Thanks for including this section

### New Web API

We considered creating a new API on `navigator` to return a `FileSystemDirectoryHandle` of the app's LocalFolder directory. This was rejected primarily because of the long term cost of maintaining a non-standard API. 

### File migration

We considered a solution that migrates the contents of the LocalFolder directory into OPFS storage. The drawbacks to this solution includes:

* It could take a long time to migrate large files. There is no way for the calling code to register a callback to wait for migration to be completed without introducing a new API, negating the advantage of not creating a new API on `navigator`.
* Migration of files could fail due the lack of storage space and other I/O errors. There is no good way to communicate such errors to the calling code without introducing a new API.

> ## Privacy And Security Considerations

Reviewer: Just in terms of fully documenting security impact, does the [storage access permissions policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Permissions-Policy/storage-access) apply to OPFS at all, in terms of being able to limit access to iframes?

Author: Storage Access API (and the `storage-access` permission policy seems to only control access to cookie storage (window.cookie). It does not seem relevant to OPFS.

# ApplicationData.LocalFolder Access From Microsoft Store PWA 

Author: [Lu Huang](https://github.com/luhuangMSFT)

## Status of this Document

This document is a starting point for engaging the community and standards bodies in developing collaborative solutions fit for standardization. As the solutions to problems described in this document progress along the standards-track, we will retain this document as an archive and use this section to keep the community up-to-date with the most current standards venue and content location of future work and discussions.

* This document status: **Active**
* Expected venue: [Microsoft Edge Explainers](https://github.com/MicrosoftEdge/MSEdgeExplainers)
* **Current version: this document**

## Introduction

> [UWP apps](https://learn.microsoft.com/en-us/windows/uwp/get-started/universal-application-platform-guide) often make use of the [`ApplicationData.LocalFolder`](https://learn.microsoft.com/en-us/uwp/api/windows.storage.applicationdata.localfolder) API for access to local storage. This API manages storage in a system file directory unique to the [Package Family Name (PFN)](https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/package-identity-overview#package-family-name) of the installed app package. A UWP app distributed by the [Microsoft Store](https://apps.microsoft.com/home) may be replaced with a PWA app under the same PFN. When this happens, existing files in the `LocalFolder` directory become inaccessible to the PWA. 

Reviewer 3: > existing files in the `LocalFolder` directory become inaccessible to the PWA

Why? What's different that affects this?

Reviewer 4: "Sandboxing" of PWA vs Native App environment does not allow the PWA to access the hard drive location of he native app it is replacing. That is the way it exists today *without* this feature. For data migration purposes, this feature would allow the PWA to access this location for the purposes of data migration.

Author: Edited to explain that this is because WinRT APIs are not exposed to the Web once a UWP app gets replaced with a PWA app package.

In this explainer, we propose a solution that allows Microsoft Store PWAs to read and delete files from the `LocalFolder` directory belonging to their PFN. This allows apps to provide a more seamless user experience after an update and to reclaim storage space.

The proposed solution works by making use of the [Origin Private File System (OPFS)](https://web.dev/articles/origin-private-file-system) and the [File System Access](https://wicg.github.io/file-system-access/) APIs. It exposes an app's `LocalFolder` file system directory as an entry in the app origin's OPFS root directory.

## Goals

* Enable a Microsoft Store PWA to:
   * read files from the `LocalFolder` directory specific to its PFN.

>    * delete any file and sub-directory in its `LocalFolder` directory. 

Reviewer: It's an explicit decision that this feature will _not_ allow _write_ access to the content of this folder (it's the first non-goal of the Explainer).  We're providing access to read the content for migration purposes, and delete for cleanup purposes, but we don't want to encourage further usage of this directory once the app is a PWA.  The web platform provides a number of various storage solutions (like OPFS that this feature is built upon) that would be under the purview of the user agent/browser, and thus be able to honor user privacy settings (like the ability to clear local storage).  This folder is outside the scope of what the user agent maintains, and thus undesirable for regular app data storage.

>    * delete any file and sub-directory in its `LocalFolder` directory. 

Author: I think the inconsistent phrasing (LocalFolder directory vs file and sub-directory in LocalFolder directory now actually works given new changes to the deletion logic. Please take another look, referencing the table of deletion actions.

* Support web apps not installed or managed by the Microsoft Store.
* Support web apps on non-Windows platforms.
* Standardization as a web specification.

## Proposed Solution 

### Existing behavior

[`navigator.storage.getDirectory()`](https://developer.mozilla.org/en-US/docs/Web/API/StorageManager/getDirectory) is an existing API that returns a promise to a [FileSystemDirectoryHandle](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemDirectoryHandle) which represents the root of directory the origin's OPFS storage space. By default, an origin's OPFS root directory has no entries.

### New behavior

When configured correctly, the directory handle for OPFS root will contain an additional entry that represents the app's `LocalFolder` directory. This entry can be retrieved as a `FileSystemDirectoryHandle` by calling [`.entries()`](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemDirectoryHandle/entries) or [`.getDirectoryHandle(...)`](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemDirectoryHandle/getDirectoryHandle) on the OPFS root directory handle.

> Like other OPFS handles, the `LocalFolder` directory handle and its contents will have [`readwrite`](https://wicg.github.io/file-system-access/#dom-filesystempermissionmode-readwrite) permission by default. Unlike other handles with `readwrite` permission, a [FileSystemFileHandle](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemFileHandle) from `LocalFolder` will always throw a [`NoModificationAllowedError`](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemFileHandle/createWritable#nomodificationallowederror) exception if [`createWritable()`](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemFileHandle/createWritable#exceptions) is used. This allows data in `LocalFolder` to be read and cleared but not created or edited.

Reviewer: Won't you need to do something similar for `FileSystemDirectoryHandle` and `getDirectoryHandle()` so that attempts to create new sub-directories result in the same exception?

Author: Called out `getDirectoryHandle()` and `getFIleHandle()` with `create===true` as well.

    if (opfsRoot) {
        let localFolder = await opfsRoot.getDirectoryHandle("microsoft_store_app_local_folder_APPID.37853FC22B2CE_6rarf9sa4v8jt", {create: false}); 
    }
```
*Example: looking for the `LocalFolder` entry under the OPFS root directory.*

*Note: the `create` option should be `false` to determine if the system `LocalFolder` is present. If it is true, `getDirectoryHandle` will always return a valid handle as it could create a directory handle of the same name in actual OPFS storage.*

```JS
    // .entries()
    let opfsRoot = await navigator.storage.getDirectory();
    if (opfsRoot) {
        for await (const [key, value] of opfsRoot.entries()) {
            if (key === 'microsoft_store_app_local_folder_APPID.37853FC22B2CE_6rarf9sa4v8jt') {
                // ...
            }
        }
    }
```
*Example: Looking for the `LocalFolder` entry by iterating entries under the OPFS root directory.*

#### Deleting LocalFolder and its contents 

`LocalFolder`'s contents can be cleared by calling `.remove()` or `.removeEntry(...)` on [FileSystemHandles](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemHandle). 

| Action | Result |
|--------|--------|

> | `.remove()` on OPFS root handle | Will not enumerate or clear LocalFolder. |

Author: No, by the logic of `.remove()` on OPFS root won't enumerate `LocalFolder`.

> | `.remove()` on OPFS root handle | Will not enumerate or clear LocalFolder. |

Author: Added clarification that `.remove()` will not enumerate or clear `LocalFolder` even if `recursive` option is used.

1. A malicious party could create executable files on the file system.

1. Third party scripts could read data in `LocalFolder` even though the app developer is not aware of this feature. 

1. Scripts in embedded iframes could try to access data in `LocalFolder`. 

### Mitigations

1. The `FileSystemHandle` design does not support walking up the directory structure.

1. The `LocalFolder` directory handle and its contents will not allow the modification of existing files or creation of new files. This prevents the creation of executable files and the creation of files outside of the LocalFolder directory. Additionally, the implementation of the File System Access API in Chromium prevents creation of some executable file types.

1. The app has to opt in explictly using the `related_applications` field in the web app manifest. The app developer must understand the risks involved and be responsible for the actions of third party scripts used. Most apps that do not intend to access `LocalFolder` will not expose it unknowingly to third parties.

1. The `LocalFolder` handle entry will not be visible to iframes as OPFS storage is partitioned by origin.

## Privacy risks

> 1. `LocalFolder` data in the file system can be accessed by a PWA through a Web API without prompting the user for permission. 

Author: The 2nd item is about the presence of the entry. Reworded to use "presence". The first item is about whether the web should access local, non-web, data without asking for user permission. Reworded for clarity.

1. Third party scripts can use the presence of `LocalFolder` data to determine if that particular Windows Store app is installed on the client machine. This information could potentially be used for fingerprinting. 

>     * Inspecting `display-mode` is an existing method to check if a site is installed locally as an app but this method does not work when the user navigates the site in a normal browesr tab.  

Author: Added.

    * The fingerprinting surface does not increase significantly as there is any existing method to determine if apps are installed. 

>     * As the installed app information would be different for each site, it cannot be easily used in a comparable fingerprint from different sites. 

Author: What I'm trying to express: 
You can determine from `a.com` that PFN=A is installed. You can determine from `b.com` that PFN=B is installed. You cannot know from `b.com` that PFN=A is installed. If you use this bit of information in a fingerprint, the fingerprint calculated on a.com is not comparable to the fingerprint calculated on b.com. It can be used to distinguish 2 clients (one installed, one not) on a.com, but it cannot easily be used to track the same client across 2 sites.

>     * As the installed app information would be different for each site, it cannot be easily used in a comparable fingerprint from different sites. 

Reviewer: I'd probably rephrase this.  It's true that you can't determine `a.com`'s installation status from `b.com`, but you can now determine "is _this_ site installed from Microsoft Store".  That additional information could be combined with other existing data that might be able to link the user across both sites (possibly a cookie) and then know which sites have been installed from Microsoft Store.

So, maybe something like this:

` * An origin is only able to use this feature to determine if it itself is installed from the Microsoft Store.  It is unable to determine the installation status of any other origin.  This information can be combined with other unrelated, existing data however to potentially increase the fingerprinting risk.`

Or something like that.

As an aside, you should update all references of `Windows Store` -> `Microsoft Store`.

## Security risks 

1. A malicious party could attempt to gain access to files outside of the boundary of the `LocalFolder` directory.

1. A malicious party could create executable files on the file system.

1. Third party scripts could read data in `LocalFolder` even though the app developer is not aware of this feature. 

1. Scripts in embedded iframes could try to access data in `LocalFolder`. 

### Mitigations

1. The `FileSystemHandle` design does not support walking up the directory structure.

1. The `LocalFolder` directory handle and its contents will not allow the modification of existing files or creation of new files. This prevents the creation of executable files and the creation of files outside of the LocalFolder directory. Additionally, the implementation of the File System Access API in Chromium prevents creation of some executable file types.

1. The app has to opt in explictly using the `related_applications` field in the web app manifest. The app developer must understand the risks involved and be responsible for the actions of third party scripts used. Most apps that do not intend to access `LocalFolder` will not expose it unknowingly to third parties.

> 1. The `LocalFolder` handle entry will not be visible to iframes as OPFS storage is partitioned by origin.

Reviewer 5: Does this need an additional qualifier about what type of iframe it won't be visible to? e.g. "will not be not visible to cross-origin iframes"?

## Privacy risks

1. `LocalFolder` data in the file system can be accessed by a PWA through a Web API without prompting the user for permission. There could be personally identifiable information (PII) contained within this data.  

> 1. Third party scripts can use the presence of a `LocalFolder` entry to determine if a particular Windows Store app is installed on the client machine. This information could potentially be used for fingerprinting. 

Reviewer 5: For clarity, can you say "Any script, including those from third-party origins that may not be owned by the application developer, that are loaded into a document that's the same origin as the top-level frame, can use the presence ..."?

And further clarify that it's not allowing determining "if a particular Windows Store app is installed" (which could be read to mean allowing querying of any store app). Instead, it's allowing determining, "if a Windows Store app associated with the origin is installed."

Author: Done.
